protocol/contracts/core/MultiOtcOffer.sol

```diff
@@ -0,0 +1,312 @@
+import { EnumerableSet } from "@openzeppelin/contracts/utils/structs/EnumerableSet.sol";
+import { TempleElevatedAccess } from "contracts/v2/access/TempleElevatedAccess.sol";
+import { IMultiOtcOffer } from "contracts/interfaces/core/IMultiOtcOffer.sol";
+
+
+/**
+ * @title Multi OTC Offer
+ *
+ * @notice Temple offers OTC purchases to users on certain tokens - slippage and price impact free.
+ * Temple sets the offer price and users can swap tokens for any arbitrary size at this price, up to some
+ * max amount of treasury funds (determined by the `fundsOwner` balance and ERC20 approvals). 
+ * This contract is set up to take care of multiple OTC markets. Elevated access address can add and remove OTC markets.
+ */
+contract MultiOtcOffer is IMultiOtcOffer, Pausable, TempleElevatedAccess {
+    using SafeERC20 for IERC20Metadata;
+    using EnumerableSet for EnumerableSet.Bytes32Set;
+
+    /// @notice The number of decimal places represented by `offerPrice`
+    uint8 public constant OFFER_PRICE_DECIMALS = 18;
+
+    /// @notice Mapping of OTC market Ids to details 
+    mapping(bytes32 marketId => OTCMarketInfo marketInfo) private otcMarketInfo;
+    /// @notice keep track of all OTC Market Ids
+    EnumerableSet.Bytes32Set private _otcMarketIds;
+    /// @notice Reverse map market Ids to buy and sell tokens
+    mapping(bytes32 marketId => address[] tokens) private marketIdToTokens;
+
+    constructor(
+        address _initialRescuer,
+        address _initialExecutor
+    )  TempleElevatedAccess(_initialRescuer, _initialExecutor) {}
+
+    /**
+     * @notice Add new OTC market.
+     * @param _otcMarketInfo OTC Market details
+     * @return marketId Bytes32 Market Id
+     */
+    function addOtcMarket(
+        OTCMarketInfo calldata _otcMarketInfo
+    ) external override onlyElevatedAccess returns (bytes32 marketId) {
+        if (_otcMarketInfo.fundsOwner == address(0) ||
+            address(_otcMarketInfo.userBuyToken) == address(0) ||
+            address(_otcMarketInfo.userSellToken) == address(0)) { revert CommonEventsAndErrors.InvalidAddress(); }
+        if (_otcMarketInfo.userBuyToken == _otcMarketInfo.userSellToken) 
+            { revert InvalidTokenPair(address(_otcMarketInfo.userBuyToken), address(_otcMarketInfo.userSellToken)); }
+        if (_otcMarketInfo.minValidOfferPrice == 0 || _otcMarketInfo.minValidOfferPrice > _otcMarketInfo.maxValidOfferPrice) 
+            { revert CommonEventsAndErrors.InvalidParam(); }
+        if (_otcMarketInfo.offerPrice < _otcMarketInfo.minValidOfferPrice || _otcMarketInfo.offerPrice > _otcMarketInfo.maxValidOfferPrice) 
+            { revert CommonEventsAndErrors.InvalidParam(); }
+
+        // check existing hash for token pair
+        marketId = _createMarketHash(address(_otcMarketInfo.userBuyToken), address(_otcMarketInfo.userSellToken));
+        if (!_otcMarketIds.add(marketId)) { revert MarketPairExists(); }
+        address[] memory tokens = new address[](2);
```

Comment:
This is only ever going to be two tokens right? Would this be better represented as a struct:
```
struct MarketTokens {
    address buyToken;
    address sellToken;
}
```

So then there's no interpretation needed by clients on what each item in the dynamic array represents?

Reply:
okay will update